### ISSUE TITLE
Units in astropy Table

### ISSUE THREAD

**Issue by paulray** [nanograv/PINT#688]
There is a strange behavior in astropy Tables.  If a Table column has a unit, and you grab the whole column, or a slice of the column, you get a Column with units; but if you grab a single value you don't get a Quantity with units:
```
In [7]: ts.table['mjd_float']                                                                                            
Out[7]: 
<Column name='mjd_float' dtype='float64' unit='d' length=62>
 53478.28587141922
 53483.27670518852
 53489.46838978793
53679.875645920525
 53679.87564536506
 53679.87564492729
               ...
 54099.70978490328
 54099.70978468499
 54099.70978449154
  54099.7097843152
54099.709784159364

In [8]: ts.table['mjd_float'][0:2]                                                                                       
Out[8]: 
<Column name='mjd_float' dtype='float64' unit='d' length=2>
53478.28587141922
53483.27670518852

In [9]: ts.table['mjd_float'][0]                                                                                         
Out[9]: 53478.28587141922
```

This is unfortunate.  I wonder if this will change in a future astropy, or if there is some better way of doing this that doesn't drop the units?

**Comment by luojing1211** [comment]
@mhvk Is this the same problem as the time object table column losing the location?

**Comment by mhvk** [comment]
No, it is an historical artifact that `Column` treats units as a "description" that may or may not be of interest. We had a very long discussion about it, but in the end it was felt this could not be changed since there was code that relied on this behaviour (quite a bit of it by the main architect of `Table`).

Instead, if you want columns with units to behave as quantities, the suggestion is to use `QTable`, in which such columns in fact *are* quantities. So, that should be the solution - and I think it is well worth trying. Ideally nothing should break... (more realistically, things will break, but hopefully it will just lead to code simplification; do ping me in case of trouble!)

p.s. Note that our table docs start with a `QTable` example as that is felt to be the right model; see https://docs.astropy.org/en/latest/table/mixin_columns.html#quantity-and-qtable
p.s.2 The one thing that `Quantity` does not allow is masking. This is a show-stopper in some cases...

**Comment by paulray** [comment]
@mhvk can I ask some advice about handling `Time()` in a table?  We currently have some columns where each entry is a `Time` object. This is necessary when the attributes of each entry could be different (e.g. location and scale). However, for some columns (perhaps TDB) we might rather have it considered to be a `Time()` object with an array of values in it since that would be much faster and more efficient than an array of objects. Is there a good way to do that in a QTable or some other way?

**Comment by demorest** [comment]
Also maybe relevant is #651 .. some subtle bug with Time in Table that appeared in astropy 4.0.1.

**Comment by mhvk** [comment]
@paulray - yes, I looked at the table at some point because of the change of behaviour of `Time` for lists of input times, and hacked a bit at trying to use pure `Time` columns throughout before realizing that it would never work for the input columns, which, as you note, have different scales. 
In principle, it can work for `tdb` if you are willing to drop the location column (since that is `None` for some input times, yet given for others, which is not possible, at least not without a placeholder for location). It would actually seem fine to drop location for the `tdb` column, since the whole point of it is that the location is removed (and you have it in a different column anyway), but there is at least one tests that checks that `location` is not removed.

@demorest - I think @luojing1211 correctly identified the fix that wasn't quite right. Not quite the same issue, I think, so probably best looked at in #651 itself.

**Comment by mhvk** [comment]
p.s. Probably best to discuss further in a separate issue!? I think this one would be more about just switching to `QTable`, which helps with the units, but makes no difference for times.
